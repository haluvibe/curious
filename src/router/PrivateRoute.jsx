//@flow
import * as React from "react";
import { connect, type Connector, type MapStateToProps } from "react-redux";
import { app } from "../firebase/rebase.config.js";
import { Redirect, Route } from "react-router-dom";
import * as actions from "../actions";
import type { UserReducer } from "../reducers/userReducer";

type OwnProps = {
  component: React.ComponentType<any>,
  exact: ?boolean,
  path: string
};

type StateProps = {
  isAuthenticated: boolean
} & OwnProps;

type DispatchProps = {
  setIsAuthenticated: Function
};

type Props = StateProps & DispatchProps;

export class PrivateRoute extends React.Component<Props> {
  authListener: ?Function = () => {};

  componentWillMount () {
    const { setIsAuthenticated } = this.props;
    this.authListener = app.auth().onAuthStateChanged(function (user) {
      if (user) {
        setIsAuthenticated(true);
      } else {
        setIsAuthenticated(false);
      }
    });
  }

  componentWillUnmount () {
    this.authListener = null;
  }

  render () {
    const {
      component: Component,
      isAuthenticated,
      setIsAuthenticated,
      ...rest
    } = this.props;
    return (
      <Route
        {...rest}
        render={() => {
          return isAuthenticated ? <Component /> : <Redirect to="/login" />;
        }}
      />
    );
  }
}

const mapStateToProps: MapStateToProps<*, *, *> = (
  state: { userReducer: UserReducer },
  ownProps
): StateProps => ({
  isAuthenticated: state.userReducer.isAuthenticated,
  exact: ownProps.exact,
  path: ownProps.path,
  component: ownProps.component
});

const mapDispatchToProps = {
  setIsAuthenticated: actions.setIsAuthenticated
};

const connector: Connector<OwnProps, Props> = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default connector(PrivateRoute);
