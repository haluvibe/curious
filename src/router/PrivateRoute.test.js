import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { PrivateRoute } from "./PrivateRoute";

const MockComponent = () => {
  <div>mock</div>;
};

const mockProps = {
  isAuthenticated: false,
  exact: true,
  path: "/form",
  component: MockComponent,
  setIsAuthenticated: () => {}
};

describe("PrivateRoute", () => {
  test("component should render as expected", () => {
    const component = shallow(
        <MemoryRouter>
          <PrivateRoute {...mockProps} />
        </MemoryRouter>
      ),
      tree = toJson(component.dive().dive());
    // expect(tree).toMatchSnapshot();
    expect(tree).toMatchSnapshot();
  });
});
