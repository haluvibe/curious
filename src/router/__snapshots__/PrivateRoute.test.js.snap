// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute component should render as expected 1`] = `
<Route
  exact={true}
  path="/form"
  render={[Function]}
/>
`;
